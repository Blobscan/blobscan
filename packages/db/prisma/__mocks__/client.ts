import { beforeEach } from "vitest";
import { mockDeep, mockReset } from "vitest-mock-extended";

import type { BlobscanPrismaClient } from "../index";

beforeEach(() => {
  mockReset(prisma);
});

const prisma = mockDeep<BlobscanPrismaClient>();

export default prisma;
