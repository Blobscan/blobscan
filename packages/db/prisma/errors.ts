export {
  PrismaClientKnownRequestError,
  PrismaClientUnknownRequestError,
  PrismaClientValidationError,
} from "@prisma/client/runtime/library";
