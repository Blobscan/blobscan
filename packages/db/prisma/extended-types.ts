import type { Prisma } from "@prisma/client";

import type { BlobscanPrismaClient } from ".";

export type ExtendedTransactionSelect = NonNullable<
  Prisma.Args<BlobscanPrismaClient["transaction"], "findFirstOrThrow">["select"]
>;

export type ExtendedBlockSelect = NonNullable<
  Prisma.Args<BlobscanPrismaClient["block"], "findFirst">["select"]
>;

export type ExtendedEthUsdPriceSelect = NonNullable<
  Prisma.Args<BlobscanPrismaClient["ethUsdPrice"], "findFirstOrThrow">["select"]
>;

export type ExtendedBlobDataStorageReferenceSelect = NonNullable<
  Prisma.Args<
    BlobscanPrismaClient["blobDataStorageReference"],
    "findFirstOrThrow"
  >["select"]
>;
