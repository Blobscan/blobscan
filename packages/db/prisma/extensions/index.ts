export * from "./base";
export * from "./stats";
