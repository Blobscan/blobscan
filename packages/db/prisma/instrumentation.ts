import { api, SemanticAttributes } from "@blobscan/open-telemetry";

export const tracer = api.trace.getTracer("@blobscan/db");

export function curryPrismaExtensionFnSpan(extensionName: string) {
  return function (modelName: string) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return function <F extends () => any>(fnName: string, fn: F) {
      return tracer.startActiveSpan<
        (span: api.Span) => Promise<Awaited<ReturnType<F>>>
      >(
        "prisma:extension",
        {
          attributes: {
            [SemanticAttributes.CODE_FUNCTION]: fnName,
            extension: extensionName,
            model: modelName,
          },
        },
        async (span) => {
          const result = await fn();

          span.end();

          return result;
        }
      );
    };
  };
}
