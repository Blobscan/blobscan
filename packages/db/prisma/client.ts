import { PrismaClient } from "@prisma/client";

import { logger } from "@blobscan/logger";

import { baseExtension, statsExtension } from "./extensions";

const globalForPrisma = globalThis as unknown as {
  prisma: PrismaClient;
};

let prisma_ = globalForPrisma.prisma;

if (!prisma_) {
  const p = new PrismaClient({
    log:
      process.env.NODE_ENV === "development"
        ? [
            { emit: "event", level: "query" },
            { emit: "event", level: "error" },
            { emit: "event", level: "warn" },
          ]
        : [
            { emit: "event", level: "warn" },
            { emit: "event", level: "error" },
          ],
  });

  // eslint-disable-next-line turbo/no-undeclared-env-vars
  if (process.env.MODE !== "test") {
    p.$on("query", (e) => {
      logger.debug(`${e.query}\nDuration=${e.duration}ms`);
    });

    p.$on("error", (e) => {
      logger.error(e.message);
    });

    p.$on("warn", (e) => {
      logger.warn(e.message);
    });
  }

  prisma_ = p;
}

export const prisma = prisma_.$extends(baseExtension).$extends(statsExtension);

if (process.env.NODE_ENV !== "production") globalForPrisma.prisma = prisma_;

export type BlobscanPrismaClient = typeof prisma;

export default prisma;
