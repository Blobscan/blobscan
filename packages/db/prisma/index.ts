import { PrismaClient } from "@prisma/client";

import { baseExtension, statsExtension } from "./extensions";

const globalForPrisma = globalThis as unknown as { prisma: PrismaClient };

const prisma_ =
  globalForPrisma.prisma ||
  new PrismaClient({
    log:
      process.env.NODE_ENV === "development"
        ? ["query", "error", "warn"]
        : ["error"],
  });

export const prisma = prisma_.$extends(baseExtension).$extends(statsExtension);

export type BlobscanPrismaClient = typeof prisma;

export * from "./types";
export * from "./utils/dates";

if (process.env.NODE_ENV !== "production") globalForPrisma.prisma = prisma_;
