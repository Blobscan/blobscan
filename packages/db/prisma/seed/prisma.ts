import { PrismaClient } from "@prisma/client";

import { baseExtension, statsExtension } from "../extensions";

export function createLocalPrisma() {
  return new PrismaClient().$extends(baseExtension).$extends(statsExtension);
}
