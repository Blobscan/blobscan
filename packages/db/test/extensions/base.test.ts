import { BlobStorage } from "@prisma/client";
import { describe, it, expect } from "vitest";

import { prisma } from "../../prisma";

describe("Base Extension", () => {
  describe("Address model", () => {
    it("should upsert addresses from transactions", async () => {
      const txs = [
        {
          from: "address8",
          to: "address1",
          blockNumber: 1002,
        },
        {
          from: "address4",
          to: "address8",
          blockNumber: 1002,
        },
        {
          from: "address3",
          to: "address4",
          blockNumber: 1002,
        },
        {
          from: "address6",
          to: "address3",
          blockNumber: 1002,
        },
      ];
      await prisma.address.upsertAddressesFromTransactions(txs);

      const [all, address1, address3, address8] = await Promise.all([
        prisma.address.findMany({
          select: {
            address: true,
            firstBlockNumberAsSender: true,
            firstBlockNumberAsReceiver: true,
          },
          orderBy: {
            address: "asc",
          },
        }),
        prisma.address.findUnique({
          select: {
            address: true,
            firstBlockNumberAsSender: true,
            firstBlockNumberAsReceiver: true,
          },
          where: {
            address: "address1",
          },
        }),
        prisma.address.findUnique({
          select: {
            address: true,
            firstBlockNumberAsSender: true,
            firstBlockNumberAsReceiver: true,
          },
          where: {
            address: "address3",
          },
        }),
        prisma.address.findUnique({
          select: {
            address: true,
            firstBlockNumberAsSender: true,
            firstBlockNumberAsReceiver: true,
          },
          where: {
            address: "address8",
          },
        }),
      ]);

      expect(all).toHaveLength(7);
      expect(all).toMatchSnapshot();

      // Existing address should create new field
      expect(address1?.firstBlockNumberAsReceiver).toBe(1002);

      // Existing address should not update fields
      expect(address3?.firstBlockNumberAsReceiver).toBe(1001);
      expect(address3?.firstBlockNumberAsSender).toBe(1001);

      // New address should create both fields
      expect(address8?.firstBlockNumberAsReceiver).toBe(1002);
      expect(address8?.firstBlockNumberAsSender).toBe(1002);
    });
  });

  describe("Blob model", () => {
    it("should filter new blobs", async () => {
      const rawBlobs = [
        {
          versionedHash: "blobHash001",
          commitment: "commitment001",
          txHash: "txHash001",
          index: 0,
          data: "data001",
        },
        {
          versionedHash: "blobHash002",
          commitment: "commitment002",
          txHash: "txHash002",
          index: 0,
          data: "data002",
        },
        {
          versionedHash: "blobHash007",
          commitment: "commitment007",
          txHash: "txHash007",
          index: 0,
          data: "data007",
        },
        {
          versionedHash: "blobHash008",
          commitment: "commitment008",
          txHash: "txHash008",
          index: 0,
          data: "data008",
        },
      ];

      const result = await prisma.blob.filterNewBlobs(rawBlobs);

      expect(result).toHaveLength(2);
      expect(result).toMatchSnapshot();
    });

    it("should upsert many blobs", async () => {
      const blobs = [
        {
          versionedHash: "blobHash005",
          commitment: "newCommitment005",
          size: 1400,
          firstBlockNumber: 1001,
        },
        {
          versionedHash: "blobHash006",
          commitment: "commitment006",
          size: 1200,
          firstBlockNumber: 1001,
        },
        {
          versionedHash: "blobHash007",
          commitment: "commitment007",
          size: 1300,
          firstBlockNumber: 1002,
        },
      ];

      await prisma.blob.upsertMany(blobs);

      const result = await prisma.blob.findMany({
        select: {
          versionedHash: true,
          commitment: true,
          size: true,
          firstBlockNumber: true,
        },
        orderBy: {
          versionedHash: "asc",
        },
      });
      expect(result).toHaveLength(8);
      expect(result).toMatchSnapshot();
    });
  });

  describe("BlobDataStorageReference model", () => {
    it("should upsert many references", async () => {
      const refs = [
        {
          blobHash: "blobHash001",
          blobStorage: BlobStorage.GOOGLE,
          dataReference: "gs://new-bucket/blobHash001",
        },
        {
          blobHash: "blobHash001",
          blobStorage: BlobStorage.SWARM,
          dataReference: "bzz://some-newhash-for-blobHash001",
        },
        {
          blobHash: "blobHash002",
          blobStorage: BlobStorage.GOOGLE,
          dataReference: "gs://new-bucket/blobHash002",
        },
      ];

      await prisma.blobDataStorageReference.upsertMany(refs);

      const result = await prisma.blobDataStorageReference.findMany({
        where: {
          blobHash: {
            in: ["blobHash001", "blobHash002"],
          },
        },
        select: {
          blobHash: true,
          blobStorage: true,
          dataReference: true,
        },
      });

      expect(result).toHaveLength(3);
      expect(result).toMatchSnapshot();
    });
  });

  describe("Block model", () => {
    it("should find the latest block", async () => {
      const result = await prisma.block.findLatest();

      expect(result).toMatchSnapshot();
    });
  });

  describe("Transaction model", () => {
    it("should upsert many transactions", async () => {
      const transactions = [
        {
          hash: "txHash001",
          fromId: "address1",
          toId: "address3",
          blockNumber: 1002,
          maxFeePerBlobGas: BigInt(100),
          gasPrice: BigInt(10),
          blobAsCalldataGasUsed: BigInt(1000),
        },
        {
          hash: "txHash003",
          fromId: "address5",
          toId: "address3",
          blockNumber: 1001,
          maxFeePerBlobGas: BigInt(120),
          gasPrice: BigInt(5),
          blobAsCalldataGasUsed: BigInt(500),
        },
        {
          hash: "txHash004",
          fromId: "address1",
          toId: "address3",
          blockNumber: 1002,
          maxFeePerBlobGas: BigInt(100),
          gasPrice: BigInt(10),
          blobAsCalldataGasUsed: BigInt(1000),
        },
        {
          hash: "txHash008",
          fromId: "address5",
          toId: "address6",
          blockNumber: 1002,
          maxFeePerBlobGas: BigInt(100),
          gasPrice: BigInt(10),
          blobAsCalldataGasUsed: BigInt(1000),
        },
      ];

      await prisma.transaction.upsertMany(transactions);

      const result = await prisma.transaction.findMany({
        select: {
          hash: true,
          fromId: true,
          toId: true,
          blockNumber: true,
          maxFeePerBlobGas: true,
          gasPrice: true,
          blobAsCalldataGasUsed: true,
        },
        orderBy: {
          hash: "asc",
        },
      });
      expect(result).toHaveLength(7);
      expect(result).toMatchSnapshot();
    });
  });
});
