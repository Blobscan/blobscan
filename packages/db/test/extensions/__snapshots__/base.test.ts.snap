// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Base Extension > Address model > should upsert addresses from transactions 1`] = `
[
  {
    "address": "address1",
    "firstBlockNumberAsReceiver": 1002,
    "firstBlockNumberAsSender": 1001,
  },
  {
    "address": "address2",
    "firstBlockNumberAsReceiver": 1001,
    "firstBlockNumberAsSender": null,
  },
  {
    "address": "address3",
    "firstBlockNumberAsReceiver": 1001,
    "firstBlockNumberAsSender": 1001,
  },
  {
    "address": "address4",
    "firstBlockNumberAsReceiver": 1001,
    "firstBlockNumberAsSender": 1002,
  },
  {
    "address": "address5",
    "firstBlockNumberAsReceiver": 1002,
    "firstBlockNumberAsSender": 1001,
  },
  {
    "address": "address6",
    "firstBlockNumberAsReceiver": 1001,
    "firstBlockNumberAsSender": 1002,
  },
  {
    "address": "address8",
    "firstBlockNumberAsReceiver": 1002,
    "firstBlockNumberAsSender": 1002,
  },
]
`;

exports[`Base Extension > Blob model > should filter new blobs 1`] = `
[
  {
    "commitment": "commitment007",
    "data": "data007",
    "txHash": "txHash007",
    "versionedHash": "blobHash007",
  },
  {
    "commitment": "commitment008",
    "data": "data008",
    "txHash": "txHash008",
    "versionedHash": "blobHash008",
  },
]
`;

exports[`Base Extension > Blob model > should upsert many blobs 1`] = `
[
  {
    "commitment": "0xb4f67eb0771fbbf1b06b88ce0e23383daf994320508d44dd30dbd507f598c0d9b3da5a152e41a0428375060c3803b983",
    "firstBlockNumber": 1002,
    "size": 1500,
    "versionedHash": "0x010001c79d78a76fb9b4bab3896ee3ea32f3e2607da7801eb1a92da39d6c1368",
  },
  {
    "commitment": "commitment001",
    "firstBlockNumber": 1001,
    "size": 1000,
    "versionedHash": "blobHash001",
  },
  {
    "commitment": "commitment002",
    "firstBlockNumber": 1001,
    "size": 1100,
    "versionedHash": "blobHash002",
  },
  {
    "commitment": "commitment003",
    "firstBlockNumber": 1001,
    "size": 1200,
    "versionedHash": "blobHash003",
  },
  {
    "commitment": "commitment004",
    "firstBlockNumber": 1002,
    "size": 1300,
    "versionedHash": "blobHash004",
  },
  {
    "commitment": "newCommitment005",
    "firstBlockNumber": 1001,
    "size": 1400,
    "versionedHash": "blobHash005",
  },
  {
    "commitment": "commitment006",
    "firstBlockNumber": 1001,
    "size": 1200,
    "versionedHash": "blobHash006",
  },
  {
    "commitment": "commitment007",
    "firstBlockNumber": 1002,
    "size": 1300,
    "versionedHash": "blobHash007",
  },
]
`;

exports[`Base Extension > BlobDataStorageReference model > should upsert many references 1`] = `
[
  {
    "blobHash": "blobHash001",
    "blobStorage": "GOOGLE",
    "dataReference": "gs://new-bucket/blobHash001",
  },
  {
    "blobHash": "blobHash001",
    "blobStorage": "SWARM",
    "dataReference": "bzz://some-newhash-for-blobHash001",
  },
  {
    "blobHash": "blobHash002",
    "blobStorage": "GOOGLE",
    "dataReference": "gs://new-bucket/blobHash002",
  },
]
`;

exports[`Base Extension > Block model > should find the latest block 1`] = `
{
  "blobAsCalldataGasUsed": 255000n,
  "blobGasPrice": 22n,
  "blobGasUsed": 5500000n,
  "excessBlobGas": 15000n,
  "hash": "blockHash002",
  "insertedAt": 2023-08-26T12:10:00.000Z,
  "number": 1002,
  "slot": 102,
  "timestamp": 2023-08-26T12:00:00.000Z,
  "updatedAt": 2023-08-26T12:10:00.000Z,
}
`;

exports[`Base Extension > Transaction model > should upsert many transactions 1`] = `
[
  {
    "blobAsCalldataGasUsed": 1000n,
    "blockNumber": 1002,
    "fromId": "address5",
    "gasPrice": 10n,
    "hash": "0x5be77167b05f39ea8950f11b0da2bdfec6e04055030068b051ac5a43aaf251e9",
    "maxFeePerBlobGas": 100n,
    "toId": "address6",
  },
  {
    "blobAsCalldataGasUsed": 1000n,
    "blockNumber": 1002,
    "fromId": "address1",
    "gasPrice": 10n,
    "hash": "txHash001",
    "maxFeePerBlobGas": 100n,
    "toId": "address3",
  },
  {
    "blobAsCalldataGasUsed": 1100n,
    "blockNumber": 1001,
    "fromId": "address3",
    "gasPrice": 11n,
    "hash": "txHash002",
    "maxFeePerBlobGas": 110n,
    "toId": "address4",
  },
  {
    "blobAsCalldataGasUsed": 500n,
    "blockNumber": 1001,
    "fromId": "address5",
    "gasPrice": 5n,
    "hash": "txHash003",
    "maxFeePerBlobGas": 120n,
    "toId": "address3",
  },
  {
    "blobAsCalldataGasUsed": 1000n,
    "blockNumber": 1002,
    "fromId": "address1",
    "gasPrice": 10n,
    "hash": "txHash004",
    "maxFeePerBlobGas": 100n,
    "toId": "address3",
  },
  {
    "blobAsCalldataGasUsed": 1000n,
    "blockNumber": 1002,
    "fromId": "address2",
    "gasPrice": 10n,
    "hash": "txHash005",
    "maxFeePerBlobGas": 100n,
    "toId": "address4",
  },
  {
    "blobAsCalldataGasUsed": 1000n,
    "blockNumber": 1002,
    "fromId": "address5",
    "gasPrice": 10n,
    "hash": "txHash008",
    "maxFeePerBlobGas": 100n,
    "toId": "address6",
  },
]
`;
